### PR TITLE
Consider non-configurable nodes in KPI computation

### PR DESCRIPTION
**Pull request by ofirgo**
Modified mixed-precision search to consider non-configurable nodes in KPI constraints.
Adding nodes that have at least one quantization configuration candidate which enables quantization (weights or activation, depending on the KPI) to the KPI computation.
Also updated kpi_data API and final config KPI accordingly.

### CODE REVIEW

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py [hunk 1]
```diff
@@ -47,23 +48,30 @@ def weights_size_kpi(mp_cfg: List[int],
...
-        node_nbits = node_qc.weights_quantization_cfg.weights_n_bits
 
-        n = _get_origin_weights_node(n)
+    if len(mp_cfg) == 0:
+        # Computing non-configurable nodes KPI
+        for n in graph.nodes:
+            if n.name not in mp_nodes and n.has_weights_quantization_enabled_candidate():
+                if len(n.candidates_quantization_cfg) == 1:
+                    node_nbits = n.candidates_quantization_cfg[0].weights_quantization_cfg.weights_n_bits
+                    node_weights_memory_in_bytes = _compute_node_weights_memory(n, node_nbits, fw_info)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:59]: Please do not pass fw_info, and use DEFAULT info in _compute_node...

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py [hunk 2]
```diff
@@ -126,35 +146,48 @@ def total_weights_activation_kpi(mp_cfg: List[int],
...
+
+    if len(mp_cfg) == 0:
+        # Computing non-configurable nodes KPI
+        for n in graph.nodes:
+            if n.name not in mp_nodes:
+                if len(n.candidates_quantization_cfg) == 1:
+                    node_weights_memory_in_bytes = 0
+                    if n.has_weights_quantization_enabled_candidate():
+                        node_weights_nbits = n.candidates_quantization_cfg[0].weights_quantization_cfg.weights_n_bits
+                        node_weights_memory_in_bytes = _compute_node_weights_memory(n, node_weights_nbits, fw_info)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:159]: Same as above (regards fw_info).

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py [hunk 3]
```diff
@@ -309,6 +347,47 @@ def _get_origin_activation_node(n: BaseNode) -> BaseNode:
     return n
 
 
+def _compute_node_weights_memory(n, node_nbits, fw_info) -> float:
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:350]: Please add type hints

File: model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py [hunk 4]
```diff
@@ -309,6 +347,47 @@ def _get_origin_activation_node(n: BaseNode) -> BaseNode:
...
+    node_num_weights_params = 0
+    for attr in fw_info.get_kernel_op_attributes(origin_node.type):
+        if attr is not None:
+            node_num_weights_params += origin_node.get_weights_by_keys(attr).flatten().shape[0]
+
+    return node_num_weights_params * node_nbits / BITS_TO_BYTES
+
+
+def _compute_node_activation_memory(n, node_nbits) -> float:
+    """
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/kpi_tools/kpi_methods.py:374]: Please add type hints

File: model_compression_toolkit/core/runner.py [hunk 5]
```diff
@@ -122,7 +124,7 @@ def core_runner(in_model: Any,
             bit_widths_config = core_config.mixed_precision_config.configuration_overwrite
 
     else:
-        bit_widths_config = None
+        bit_widths_config = []
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/runner.py:127]: Doesn't set_bit_widths (in line 129) expect None if a single precision is used?

File: model_compression_toolkit/core/runner.py [hunk 6]
```diff
@@ -132,20 +134,13 @@ def core_runner(in_model: Any,
...
 
-    common.Logger.info(f'Approximated model size (in bytes): {tg.get_memory()}')
-    common.Logger.info(f'Approximated compression ratio: {round(graph.get_float_memory() / (tg.get_memory() + 1e-8), 3)}')
-
+    _set_final_kpi(graph=tg,
+                   final_bit_widths_config=bit_widths_config,
+                   kpi_functions_dict=kpi_functions_mapping,
+                   fw_info=fw_info,
+                   fw_impl=fw_impl)
+    print(tg.user_info.final_kpi)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/runner.py:142]: Please remove print

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/runner.py (hunk 5 above)
Comment (ofirgo) [reply model_compression_toolkit/core/runner.py:127]: I don't think so. It uses a flag "mixed_precision_enable" to decide whether to use the given bit-width config or not

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**